stl/inc/bit

```diff
@@ -94,18 +96,117 @@ _NODISCARD constexpr _Ty rotr(const _Ty _Val, const int _Rotation) noexcept {
+// Used at compile time and when said instructions are not supported.
+template <class _Ty>
+_NODISCARD constexpr int _Popcount_helper(_Ty _Val) noexcept {
+    constexpr int _Digits = numeric_limits<_Ty>::digits;
+    // we static_cast these bit patterns in order to truncate them to the correct size
+    _Val = static_cast<_Ty>(_Val - ((_Val >> 1) & static_cast<_Ty>(0x5555'5555'5555'5555ull)));
+    _Val = static_cast<_Ty>((_Val & static_cast<_Ty>(0x3333'3333'3333'3333ull))
+                            + ((_Val >> 2) & static_cast<_Ty>(0x3333'3333'3333'3333ull)));
+    _Val = static_cast<_Ty>((_Val + (_Val >> 4)) & static_cast<_Ty>(0x0F0F'0F0F'0F0F'0F0Full));
+    for (int _ShiftDigits = 8; _ShiftDigits < _Digits; _ShiftDigits <<= 1) {
+        _Val = static_cast<_Ty>(_Val + static_cast<_Ty>(_Val >> _ShiftDigits));
+    }
+    // we want the bottom "slot" that's big enough to store _Digits
+    return static_cast<int>(_Val & static_cast<_Ty>(_Digits + _Digits - 1));
+}
+
+#if defined(_M_IX86) || defined(_M_X64)
+
+extern "C" {
+__MACHINEX86_X64(unsigned int __lzcnt(unsigned int))
+__MACHINEX86_X64(unsigned short __lzcnt16(unsigned short))
+__MACHINEX64(unsigned __int64 __lzcnt64(unsigned __int64))
+__MACHINEX86_X64(unsigned int __popcnt(unsigned int))
+__MACHINEX86_X64(unsigned short __popcnt16(unsigned short))
+__MACHINEX64(unsigned __int64 __popcnt64(unsigned __int64))
+extern int __isa_available;
+}
+
+template <class _Ty>
+_NODISCARD int _Checked_x86_x64_countl_zero(const _Ty _Val) noexcept {
+    const bool _Have_lzcnt = __isa_available >= __ISA_AVAILABLE_AVX2;
+    constexpr int _Digits  = numeric_limits<_Ty>::digits;
+    // lzcnt (when it doesn't fall back to bsr) is defined correctly for zero
+    // bsr has undefined output for zero
+    if (!_Have_lzcnt && _Val == 0) {
+        return _Digits;
+    }
+    // We use lzcnt (actually bsr if lzcnt is not supported) now that we know
+    // we're not zero. We can do this because lzcnt and bsr share the same instruction
+    // encoding.
+    if constexpr (_Digits <= 16) {
+        return static_cast<int>(__lzcnt16(_Val) - (16 - _Digits));
+    } else if constexpr (_Digits == 32) {
+        return static_cast<int>(__lzcnt(_Val));
+    } else {
+#ifdef _M_IX86
+        const unsigned int _High = _Val >> 32;
+        const auto _Low          = static_cast<unsigned int>(_Val);
+        int _Result              = _Checked_x86_x64_countl_zero(_High);
+        _Result += _High == 0 ? _Checked_x86_x64_countl_zero(_Low) : 0;
+        return _Result;
+#else // ^^^ _M_IX86 / !_M_IX86 vvv
+        return static_cast<int>(__lzcnt64(_Val));
+#endif // _M_IX86
+    }
+    // note: we don't need to call a fallback here because
+    // all supported x86 processors at least have bsr/bsf
+}
+
+template <class _Ty>
+_NODISCARD int _Checked_x86_x64_popcount(const _Ty _Val) noexcept {
+    const bool _Have_popcnt = __isa_available >= __ISA_AVAILABLE_SSE42;

```

Comment:
Similar comment for preprocessor restructuring (with a different logic change because of how `if (_Have_popcnt)` is tested).

In this case, `__ISA_AVAILABLE_SSE42` is less than `__ISA_AVAILABLE_AVX`, so the `__AVX__` predefined macro can be checked.